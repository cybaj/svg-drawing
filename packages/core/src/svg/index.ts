export * from './svg'
export * from './command'
export * from './point'
export * from './path'
