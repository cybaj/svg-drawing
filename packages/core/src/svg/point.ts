import type { PointObject } from '../types'

/**
 * Maximum number of decimal places is 5
 *
 * @todo Changeable parameter
 */
const coeffecient = 100000
const roundNumber = (num: number) => Math.round(num * coeffecient) / coeffecient

export class Point implements PointObject {
  private _x: number
  private _y: number
  constructor(x: number, y: number) {
    this._x = roundNumber(x)
    this._y = roundNumber(y)
  }

  public get x(): number {
    return this._x
  }

  public get y(): number {
    return this._y
  }

  public get degrees(): number {
    return Math.atan2(this._y, this._x) * (180 / Math.PI)
  }

  public get absoluteValue(): number {
    return Math.sqrt(Math.pow(this._x, 2.0) + Math.pow(this._y, 2.0))
  }

  public scale(r: number): Point {
    return new Point(this._x * r, this._y * r)
  }

  public scaleX(r: number): Point {
    return new Point(this._x * r, this._y)
  }

  public scaleY(r: number): Point {
    return new Point(this._x, this._y * r)
  }

  public add(p: PointObject): Point {
    return new Point(this._x + p.x, this._y + p.y)
  }

  public sub(p: PointObject): Point {
    return new Point(this._x - p.x, this._y - p.y)
  }

  public translate(p: PointObject): void {
    this._x += p.x
    this._y += p.y
  }

  public eql(p: PointObject): boolean {
    return this._x === p.x && this._y === p.y
  }

  public clone(): Point {
    return new Point(this._x, this._y)
  }

  public toJson(): PointObject {
    return {
      x: this._x,
      y: this._y,
    }
  }

  public static fromVector({
    value,
    degrees,
  }: {
    value: number
    degrees: number
  }) {
    const theta = (degrees * Math.PI) / 180
    return new Point(Math.cos(theta) * value, Math.sin(theta) * value)
  }
}
