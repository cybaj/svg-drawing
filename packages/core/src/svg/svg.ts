import type { SvgObject, SvgOption } from '../types'
import type { Path } from './path'

export class Svg {
  public paths: Path[]
  public width: number
  public height: number
  public background?: string

  constructor({ width, height, background }: SvgOption) {
    this.paths = []
    this.width = width
    this.height = height
    this.background = background
  }

  /** @todo Check height */
  public resize({ width, height }: { width: number; height: number }) {
    this.scale(width / this.width)
    this.width = width
    this.height = height
  }

  public scale(r: number): this {
    if (r !== 1) {
      for (let i = 0; i < this.paths.length; i += 1) {
        this.paths[i].scale(r)
      }
    }
    return this
  }

  public scaleX(r: number): this {
    if (r !== 1) {
      for (let i = 0; i < this.paths.length; i += 1) {
        this.paths[i].scaleX(r)
      }
    }
    return this
  }

  public scaleY(r: number): this {
    if (r !== 1) {
      for (let i = 0; i < this.paths.length; i += 1) {
        this.paths[i].scaleY(r)
      }
    }
    return this
  }

  public addPath(pa: Path | Path[]): this {
    if (Array.isArray(pa)) {
      this.paths.push(...pa)
    } else {
      this.paths.push(pa)
    }
    return this
  }

  public deletePath(deletePath: Path): this {
    this.paths = this.paths.filter((path) => path.key !== deletePath.key)
    return this
  }

  public clonePaths(): Path[] {
    return this.paths.map((p) => p.clone())
  }

  public toJson(): SvgObject {
    return {
      width: this.width,
      height: this.height,
      background: this.background,
      paths: this.paths.map((p) => p.toJson()),
    }
  }

  public copy(svg: any extends Svg ? Svg : never): this {
    this.paths = svg.clonePaths()
    if (svg.width && this.width) {
      this.scale(this.width / svg.width)
    }
    return this
  }

  public clone(): Svg {
    const svg = new Svg({
      width: this.width,
      height: this.height,
      background: this.background,
    })
    svg.addPath(this.clonePaths())
    return svg
  }
}
