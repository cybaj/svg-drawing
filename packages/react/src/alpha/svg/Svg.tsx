import React from 'react'
import type { SvgObject } from '@svg-drawing/core'
import type { HTMLAttributes } from 'react'

export const Svg = ({
  background,
  paths,
  width,
  height,
  ...rest
}: SvgObject & HTMLAttributes<SVGSVGElement>) => (
  <svg width={width} height={height} {...rest}>
    {background && <rect width={width} height={height} fill={background} />}
    {paths.map((pathAttr, i) => (
      <path key={i} {...pathAttr} />
    ))}
  </svg>
)
