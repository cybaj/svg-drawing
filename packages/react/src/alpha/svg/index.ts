export { Svg } from './Svg'
export { useSvg } from './useSvg'
export type { UseSvgOptions, UseSvg } from './types'
