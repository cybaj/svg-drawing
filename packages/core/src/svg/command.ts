import { Point } from './point'
import { roundUp } from '../utils'
import type {
  Command,
  CommandClass,
  CommandType,
  OtherCommandType,
  PointObject,
} from '../types'

/** @deprecated */
export class OtherCommand<T extends OtherCommandType> implements Command<T> {
  static Types = {
    /**
     * Horizontal
     *
     * `H 10`
     */
    H: 'H',
    /**
     * Horizontal relative
     *
     * `h 10`
     */
    h: 'h',
    /**
     * Vertical
     *
     * V 10
     */
    V: 'V',
    /**
     * Vertical relative
     *
     * `v 10`
     */
    v: 'v',
    /**
     * Arc curve
     *
     * `A 6 4 10 0 1 14 10`
     */
    A: 'A',
    /**
     * Arc curve relative
     *
     * `a 6 4 10 0 1 14 10`
     */
    a: 'a',
  } as const

  // TODO: Convert data format to number array.
  constructor(public type: T, public values: number[] = []) {}

  public set points(p: Point[]) {
    if (!p) return
    this.values = p.reduce((acc: number[], p) => [...acc, p.x, p.y], [])
  }

  public get points(): Point[] {
    return this.values.reduce(
      (acc: Point[], _, i: number) =>
        i % 2 ? acc : [...acc, new Point(this.values[i - 1], this.values[i])],
      []
    )
  }

  public get point(): Point | undefined {
    return this.points[this.points.length - 1]
  }

  public toString(): string {
    return `${this.type}${this.values.map((v) => roundUp(v)).join(' ')}`
  }

  public scale(r: number) {
    const upd = new OtherCommand(
      this.type,
      this.values.map((p) => p * r)
    )
    return upd
  }

  public scaleX(r: number) {
    return new OtherCommand(
      this.type,
      this.points.reduce((res: number[], po: Point) => {
        const upd = po.scale(r)
        return [...res, upd.x, upd.y]
      }, [])
    )
  }

  public scaleY(r: number) {
    return new OtherCommand(
      this.type,
      this.points.reduce((res: number[], po: Point) => {
        const upd = po.scaleY(r)
        return [...res, upd.x, upd.y]
      }, [])
    )
  }

  public clone() {
    return new OtherCommand(this.type, this.values.slice())
  }

  public translate(p: PointObject) {
    if (OtherCommand.Types.a === this.type) return
    this.points.map((po) => {
      po.translate(p)
    })
  }

  public static validTypes(t: any): t is OtherCommandType {
    return Object.values(OtherCommand.Types).some((arg) => arg === t)
  }
}

/**
 * Move relative
 *
 * `m 0 0`
 */
export class RelativeMove implements CommandClass<'m'> {
  public readonly type = 'm'
  public readonly relative = false
  public points: [Point]
  constructor(point: Point) {
    this.points = [point]
  }

  public get values(): number[] {
    return this.points.reduce((acc: number[], po) => [...acc, po.x, po.y], [])
  }

  public get point(): Point {
    return this.points[0]
  }

  public toString() {
    return `${this.type}${this.values.join(' ')}`
  }

  public translate(p: PointObject) {
    return
  }

  public scale(r: number) {
    return new RelativeMove(this.points[0].scale(r))
  }

  public scaleX(r: number) {
    return new RelativeMove(this.points[0].scaleX(r))
  }

  public scaleY(r: number) {
    return new RelativeMove(this.points[0].scaleY(r))
  }

  public clone() {
    return new RelativeMove(this.points[0].clone())
  }

  public toAbsolute(base: Point): Move {
    return new Move(this.points[0].add(base))
  }
}

/**
 * Move
 *
 * `M 0 0`
 */
export class Move implements CommandClass<'M'> {
  public readonly type = 'M'
  public points: [Point]
  constructor(point: Point) {
    this.points = [point]
  }

  public get values(): number[] {
    return this.points.reduce((acc: number[], po) => [...acc, po.x, po.y], [])
  }

  public get point(): Point {
    return this.points[0]
  }

  public toString() {
    return `${this.type}${this.values.join(' ')}`
  }

  public translate(p: PointObject) {
    this.points = this.points.map((po) => po.add(p)) as [Point]
  }

  public scale(r: number) {
    return new Move(this.points[0].scale(r))
  }

  public scaleX(r: number) {
    return new Move(this.points[0].scaleX(r))
  }

  public scaleY(r: number) {
    return new Move(this.points[0].scaleY(r))
  }

  public clone() {
    return new Move(this.points[0].clone())
  }

  public toRelative(base: Point): RelativeMove {
    return new RelativeMove(this.points[0].sub(base))
  }
}

/**
 * Line relative
 *
 * `l 0 0`
 */
export class RelativeLine implements CommandClass<'l'> {
  public readonly type = 'l'
  public points: [Point]
  constructor(point: Point) {
    this.points = [point]
  }

  public get values(): number[] {
    return this.points.reduce((acc: number[], po) => [...acc, po.x, po.y], [])
  }

  public get point(): Point {
    return this.points[0]
  }

  public toString() {
    return `${this.type}${this.values.join(' ')}`
  }

  public translate(p: PointObject) {
    return
  }

  public scale(r: number) {
    return new RelativeLine(this.points[0].scale(r))
  }

  public scaleX(r: number) {
    return new RelativeLine(this.points[0].scaleX(r))
  }

  public scaleY(r: number) {
    return new RelativeLine(this.points[0].scaleY(r))
  }

  public clone() {
    return new RelativeLine(this.points[0].clone())
  }

  public toAbsolute(base: Point): CommandClass<'L'> {
    return new Line(this.points[0].add(base))
  }
}

/**
 * Line
 *
 * `L 0 0`
 */
export class Line implements CommandClass<'L'> {
  public readonly type = 'L'
  public points: [Point]
  constructor(point: Point) {
    this.points = [point]
  }

  public get values(): number[] {
    return this.points.reduce((acc: number[], po) => [...acc, po.x, po.y], [])
  }

  public get point(): Point {
    return this.points[0]
  }

  public toString() {
    return `${this.type}${this.values.join(' ')}`
  }

  public translate(p: PointObject) {
    this.points = this.points.map((po) => po.add(p)) as [Point]
  }

  public scale(r: number) {
    return new Line(this.points[0].scale(r))
  }

  public scaleX(r: number) {
    return new Line(this.points[0].scaleX(r))
  }

  public scaleY(r: number) {
    return new Line(this.points[0].scaleY(r))
  }

  public clone() {
    return new Line(this.points[0].clone())
  }

  public toRelative(base: Point) {
    return new RelativeLine(this.points[0].sub(base))
  }
}

/**
 * Cubic bezier curve relative
 *
 * `c 1 1 2 2 3 3`
 */
export class RelativeCurve implements CommandClass<'c'> {
  public readonly type = 'c'
  constructor(public points: [Point, Point, Point]) {}

  public get values(): number[] {
    return this.points.reduce((acc: number[], po) => [...acc, po.x, po.y], [])
  }

  public get point(): Point {
    return this.points[2]
  }

  public toString() {
    return `${this.type}${this.values.join(' ')}`
  }

  public translate(p: PointObject) {
    return
  }

  public scale(r: number) {
    return new RelativeCurve(
      this.points.map((p) => p.scaleX(r)) as [Point, Point, Point]
    )
  }

  public scaleX(r: number) {
    return new RelativeCurve(
      this.points.map((p) => p.scaleX(r)) as [Point, Point, Point]
    )
  }

  public scaleY(r: number) {
    return new RelativeCurve(
      this.points.map((p) => p.scaleY(r)) as [Point, Point, Point]
    )
  }

  public clone() {
    return new RelativeCurve(
      this.points.map((po) => po.clone()) as [Point, Point, Point]
    )
  }

  public toAbsolute(base: Point) {
    return new Curve(
      this.points.map((po) => po.add(base)) as [Point, Point, Point]
    )
  }
}

/**
 * Cubic bezier curve
 *
 * `C 1 1 2 2 3 3`
 */
export class Curve implements CommandClass<'C'> {
  public readonly type = 'C'
  constructor(public points: [Point, Point, Point]) {}

  public get values(): [number, number, number, number, number, number] {
    return this.points.reduce(
      (acc: number[], po) => [...acc, po.x, po.y],
      []
    ) as [number, number, number, number, number, number]
  }

  public get point(): Point {
    return this.points[2]
  }

  public toString() {
    return `${this.type}${this.values.join(' ')}`
  }

  public translate(p: PointObject) {
    this.points = this.points.map((po) => po.add(p)) as [Point, Point, Point]
  }

  public scale(r: number) {
    return new Curve(
      this.points.map((p) => p.scale(r)) as [Point, Point, Point]
    )
  }

  public scaleX(r: number) {
    return new Curve(
      this.points.map((p) => p.scaleX(r)) as [Point, Point, Point]
    )
  }

  public scaleY(r: number) {
    return new Curve(
      this.points.map((p) => p.scaleY(r)) as [Point, Point, Point]
    )
  }

  public clone() {
    return new Curve(this.points.map((p) => p.clone()) as [Point, Point, Point])
  }

  public toRelative(base: Point) {
    return new RelativeCurve(
      this.points.map((po) => po.sub(base)) as [Point, Point, Point]
    )
  }
}

/**
 * Shortcut curve
 *
 * `S 10 60 10 30`
 */
export class ShortcutCurve implements CommandClass<'S'> {
  public readonly type = 'S'
  constructor(public points: [Point, Point]) {}

  public get values(): [number, number, number, number] {
    return this.points.reduce(
      (acc: number[], po) => [...acc, po.x, po.y],
      []
    ) as [number, number, number, number]
  }

  public get point(): Point {
    return this.points[1]
  }

  public toString() {
    return `${this.type}${this.values.join(' ')}`
  }

  public translate(p: PointObject) {
    this.points = this.points.map((po) => po.add(p)) as [Point, Point]
  }

  public scale(r: number) {
    return new ShortcutCurve(
      this.points.map((p) => p.scale(r)) as [Point, Point]
    )
  }

  public scaleX(r: number) {
    return new ShortcutCurve(
      this.points.map((p) => p.scaleX(r)) as [Point, Point]
    )
  }

  public scaleY(r: number) {
    return new ShortcutCurve(
      this.points.map((p) => p.scaleY(r)) as [Point, Point]
    )
  }

  public clone() {
    return new ShortcutCurve(
      this.points.map((p) => p.clone()) as [Point, Point]
    )
  }

  public toRelative(base: Point) {
    return new RelativeShortcutCurve(
      this.points.map((po) => po.sub(base)) as [Point, Point]
    )
  }
}

/**
 * Shortcut curve relative
 *
 * `s 10 60 10 30`
 */
export class RelativeShortcutCurve implements CommandClass<'s'> {
  public readonly type = 's'
  constructor(public points: [Point, Point]) {}

  public get values(): [number, number, number, number] {
    return this.points.reduce(
      (acc: number[], po) => [...acc, po.x, po.y],
      []
    ) as [number, number, number, number]
  }

  public get point(): Point {
    return this.points[1]
  }

  public toString() {
    return `${this.type}${this.values.join(' ')}`
  }

  public translate(p: PointObject) {
    return
  }

  public scale(r: number) {
    return new RelativeShortcutCurve(
      this.points.map((p) => p.scaleX(r)) as [Point, Point]
    )
  }

  public scaleX(r: number) {
    return new RelativeShortcutCurve(
      this.points.map((p) => p.scaleX(r)) as [Point, Point]
    )
  }

  public scaleY(r: number) {
    return new RelativeShortcutCurve(
      this.points.map((p) => p.scaleY(r)) as [Point, Point]
    )
  }

  public clone() {
    return new RelativeShortcutCurve(
      this.points.map((po) => po.clone()) as [Point, Point]
    )
  }

  public toAbsolute(base: Point) {
    return new ShortcutCurve(
      this.points.map((po) => po.add(base)) as [Point, Point]
    )
  }
}

/**
 * Quadratic curve
 *
 * Q 10 60 10 30
 */

export class QuadraticCurve implements CommandClass<'Q'> {
  public readonly type = 'Q'
  constructor(public points: [Point, Point]) {}

  public get values(): [number, number, number, number] {
    return this.points.reduce(
      (acc: number[], po) => [...acc, po.x, po.y],
      []
    ) as [number, number, number, number]
  }

  public get point(): Point {
    return this.points[1]
  }

  public toString() {
    return `${this.type}${this.values.join(' ')}`
  }

  public translate(p: PointObject) {
    this.points = this.points.map((po) => po.add(p)) as [Point, Point]
  }

  public scale(r: number) {
    return new QuadraticCurve(
      this.points.map((p) => p.scale(r)) as [Point, Point]
    )
  }

  public scaleX(r: number) {
    return new QuadraticCurve(
      this.points.map((p) => p.scaleX(r)) as [Point, Point]
    )
  }

  public scaleY(r: number) {
    return new QuadraticCurve(
      this.points.map((p) => p.scaleY(r)) as [Point, Point]
    )
  }

  public clone() {
    return new QuadraticCurve(
      this.points.map((p) => p.clone()) as [Point, Point]
    )
  }

  public toRelative(base: Point) {
    return new RelativeQuadraticCurve(
      this.points.map((po) => po.sub(base)) as [Point, Point]
    )
  }
}

/**
 * Quadratic curve relative
 *
 * `q 10 60 10 30`
 */
export class RelativeQuadraticCurve implements CommandClass<'q'> {
  public readonly type = 'q'
  constructor(public points: [Point, Point]) {}

  public get values(): [number, number, number, number] {
    return this.points.reduce(
      (acc: number[], po) => [...acc, po.x, po.y],
      []
    ) as [number, number, number, number]
  }

  public get point(): Point {
    return this.points[1]
  }

  public toString() {
    return `${this.type}${this.values.join(' ')}`
  }

  public translate(p: PointObject) {
    return
  }

  public scale(r: number) {
    return new RelativeQuadraticCurve(
      this.points.map((p) => p.scaleX(r)) as [Point, Point]
    )
  }

  public scaleX(r: number) {
    return new RelativeQuadraticCurve(
      this.points.map((p) => p.scaleX(r)) as [Point, Point]
    )
  }

  public scaleY(r: number) {
    return new RelativeQuadraticCurve(
      this.points.map((p) => p.scaleY(r)) as [Point, Point]
    )
  }

  public clone() {
    return new RelativeQuadraticCurve(
      this.points.map((po) => po.clone()) as [Point, Point]
    )
  }

  public toAbsolute(base: Point) {
    return new QuadraticCurve(
      this.points.map((po) => po.add(base)) as [Point, Point]
    )
  }
}

/**
 * Close
 *
 * 'z'
 */
export class Close implements CommandClass<'z'> {
  public readonly type = 'z'

  public get values(): [] {
    return []
  }

  public get points(): [] {
    return []
  }

  public get point(): undefined {
    return undefined
  }

  public toString() {
    return `${this.type}`
  }

  public translate(_p: PointObject) {
    return
  }

  public scale(_r: number) {
    return new Close()
  }

  public scaleX(r: number) {
    return new Close()
  }

  public scaleY(r: number) {
    return new Close()
  }

  public clone() {
    return new Close()
  }
}

export const createCommand = (
  type: CommandType,
  values: number[] = []
): Command => {
  switch (type) {
    case 'M': {
      return new Move(new Point(values[0], values[1]))
    }
    case 'm': {
      return new RelativeMove(new Point(values[0], values[1]))
    }
    case 'L': {
      return new Line(new Point(values[0], values[1]))
    }
    case 'l': {
      return new RelativeLine(new Point(values[0], values[1]))
    }
    case 'C': {
      return new Curve([
        new Point(values[0], values[1]),
        new Point(values[2], values[3]),
        new Point(values[4], values[5]),
      ])
    }
    case 'c': {
      return new RelativeCurve([
        new Point(values[0], values[1]),
        new Point(values[2], values[3]),
        new Point(values[4], values[5]),
      ])
    }
    case 'Q': {
      return new QuadraticCurve([
        new Point(values[0], values[1]),
        new Point(values[2], values[3]),
      ])
    }
    case 'q': {
      return new RelativeQuadraticCurve([
        new Point(values[0], values[1]),
        new Point(values[2], values[3]),
      ])
    }
    case 'S': {
      return new ShortcutCurve([
        new Point(values[0], values[1]),
        new Point(values[2], values[3]),
      ])
    }
    case 's': {
      return new RelativeShortcutCurve([
        new Point(values[0], values[1]),
        new Point(values[2], values[3]),
      ])
    }
    case 'Z':
    case 'z': {
      return new Close()
    }

    default: {
      return new OtherCommand(type, values)
    }
  }
}
