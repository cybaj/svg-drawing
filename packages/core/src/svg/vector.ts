import { Point } from './point'

export class Vector {
  public angle: number
  public value: number
  constructor(v: number, a: number) {
    this.value = v
    this.angle = a
  }

  public toPoint(): Point {
    const x = Math.cos(this.angle) * this.value
    const y = Math.sin(this.angle) * this.value
    return new Point(x, y)
  }

  public scale(r: number): Vector {
    return new Vector(this.value * r, this.angle)
  }
}
