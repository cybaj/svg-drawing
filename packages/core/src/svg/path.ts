import { CommandType, PathObject, PointObject } from '../types'
import { kebab2camel } from '../utils'
import { Command, COMMAND_TYPE } from './command'

/**
 * TODO: refactor command.
 * The following commands are not supported. Cannot support commands that use `M` or` z` more than once
 * `M 0 0 L 1 1 Z M 1 1 L 2 2 Z`
 */
export class Path {
  public attrs: PathObject
  public commands: Command[]

  constructor({ d, ...attrs }: PathObject = {}) {
    this.attrs = attrs
    this.commands = []
    if (d) this.parseCommandString(d)
  }

  public scale(r: number): this {
    this.commands = this.commands.map((c: Command) => c.scale(r))
    this.attrs.strokeWidth = String(r * +(this.attrs.strokeWidth || 0))
    return this
  }

  public scaleX(r: number): this {
    this.commands = this.commands.map((c: Command) => c.scaleX(r))
    return this
  }

  public scaleY(r: number): this {
    this.commands = this.commands.map((c: Command) => c.scaleY(r))
    return this
  }

  public addCommand(param: Command | Command[]): this {
    if (Array.isArray(param)) {
      this.commands.push(...param)
    } else {
      this.commands.push(param)
    }
    return this
  }

  public deleteCommand(i: number): this {
    this.commands.splice(i, 1)
    return this
  }

  public getCommandString(): string {
    if (this.commands.length === 0) return ''
    return this.commands
      .map((com: Command, _i: number) => com.toString())
      .join(' ')
      .trim()
  }

  // TODO: Valid Command type
  public parseCommandString(d: string): void {
    this.commands = []
    let type: CommandType | null = null
    let value: number[] = []
    const c = d.split(' ')
    const checkType = (c: any): CommandType | null =>
      Object.values(COMMAND_TYPE).includes(c) ? c : null
    for (let i = 0; i < c.length; i += 1) {
      const t = checkType(c[i])
      // COMMAND Parse
      if (t) {
        if (!type) {
          type = t
          continue
        }
        this.commands.push(new Command(type, value))
        type = t
        value = []
        continue
      }
      if (isNaN(+c[i])) {
        return
      }
      value.push(+c[i])
    }

    if (type !== null) {
      this.commands.push(new Command(type, value))
    }
  }

  public parsePathElement(pEl: SVGPathElement): this {
    for (let i = 0; i < pEl.attributes.length; i += 1) {
      const attr: Attr | null = pEl.attributes.item(i)
      if (!attr || !attr.value) continue
      if (attr.name === 'd') {
        this.parseCommandString(attr.value)
        continue
      }
      this.attrs = {
        ...this.attrs,
        [kebab2camel(attr.name)]: attr.value,
      }
    }
    return this
  }

  public toJson(): PathObject {
    return {
      ...this.attrs,
      d: this.getCommandString(),
    }
  }

  public translate(po: PointObject): void {
    for (let i = 0; i < this.commands.length; i += 1) {
      this.commands[i].translate(po)
    }
  }

  public clone(): Path {
    const path = new Path(this.attrs)
    this.commands.map((c) => {
      path.commands.push(c.clone())
    })
    return path
  }
}
