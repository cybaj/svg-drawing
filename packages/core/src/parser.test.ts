import { parseCommandString, parseSVGString, parseSVGElement } from './parser'
import { Path } from './svg/path'

describe('parser', () => {
  it('parseCommandString', () => {
    const path = new Path()
    const testData =
      'M0 0 L1 1 C2 2 2 4 6 0 Q0 0 1 1 m0 0 l1 1 c1 1 2 2 3 3 H10 V20 h10 v20 A6 4 10 0 1 14 10 a6 4 10 0 1 14 10'
    path.addCommand(parseCommandString(testData))

    expect(path.getCommandString()).toBe(testData)
  })
  it('parseSVGString', () => {
    expect(
      parseSVGString(`
      <svg width="200" height="200">
        <path fill="#f00" stroke="#00f" stroke-width="4" d="M 1 1 L 2 2 C 3 3 5 3 7 3 Z"></path>
      </svg>`)
    ).toMatchSnapshot()
  })
  it('parseSVGElement', () => {
    const svgEl = document.createElementNS('http://www.w3.org/2000/svg', 'svg')
    svgEl.setAttribute('width', '200')
    svgEl.setAttribute('height', '200')
    const pathEl = document.createElementNS(
      'http://www.w3.org/2000/svg',
      'path'
    )
    pathEl.setAttribute('fill', '#f00')
    pathEl.setAttribute('stroke', '#00f')
    pathEl.setAttribute('stroke-width', '4')
    pathEl.setAttribute('d', 'M 1 1 L 2 2 C 3 3 5 3 7 3 Z')
    svgEl.appendChild(pathEl)
    expect(parseSVGElement(svgEl)).toMatchSnapshot()
  })
})
