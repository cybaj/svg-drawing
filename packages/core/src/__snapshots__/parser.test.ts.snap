// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`parser parseSVGElement 1`] = `
Svg {
  "background": undefined,
  "height": 200,
  "paths": Array [
    Path {
      "attrs": Object {
        "fill": "#f00",
        "stroke": "#00f",
        "strokeWidth": "4",
      },
      "commands": Array [
        Move {
          "points": Array [
            Point {
              "_x": 1,
              "_y": 1,
            },
          ],
          "type": "M",
        },
        Line {
          "points": Array [
            Point {
              "_x": 2,
              "_y": 2,
            },
          ],
          "type": "L",
        },
        Curve {
          "points": Array [
            Point {
              "_x": 3,
              "_y": 3,
            },
            Point {
              "_x": 5,
              "_y": 3,
            },
            Point {
              "_x": 7,
              "_y": 3,
            },
          ],
          "type": "C",
        },
        Close {
          "type": "z",
        },
      ],
    },
  ],
  "width": 200,
}
`;

exports[`parser parseSVGString 1`] = `
Svg {
  "background": undefined,
  "height": 200,
  "paths": Array [
    Path {
      "attrs": Object {
        "fill": "#f00",
        "stroke": "#00f",
        "strokeWidth": "4",
      },
      "commands": Array [
        Move {
          "points": Array [
            Point {
              "_x": 1,
              "_y": 1,
            },
          ],
          "type": "M",
        },
        Line {
          "points": Array [
            Point {
              "_x": 2,
              "_y": 2,
            },
          ],
          "type": "L",
        },
        Curve {
          "points": Array [
            Point {
              "_x": 3,
              "_y": 3,
            },
            Point {
              "_x": 5,
              "_y": 3,
            },
            Point {
              "_x": 7,
              "_y": 3,
            },
          ],
          "type": "C",
        },
        Close {
          "type": "z",
        },
      ],
    },
  ],
  "width": 200,
}
`;
